### PR TITLE
feat(mongodb): Add type definition for projection and sort

### PR DESCRIPTION
**Pull request by Celend**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Test the change in your own code. (Compile and run.)
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## mongodb ([<kbd>unpkg</kbd>](https://unpkg.com/browse/mongodb@latest/))
was missing the following properties:
1. The declaration doesn't match the JavaScript module 'mongodb'. Reason:
The JavaScript module can be called or constructed, but the declaration module cannot.

The most common way to resolve this error is to use 'export =' syntax.
To learn more about 'export =' syntax, see https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require.
1. MongoTimeoutError
1. MongoServerSelectionError
1. MongoWriteConcernError
1. MongoBulkWriteError

<details>
<summary>as well as these 11 other properties...</summary>
<p>BulkWriteError, Admin, Collection, GridStore, Chunk, CoreServer, CoreConnection, Map, Symbol, BSONRegExp, instrument</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 587bd5811cab6ceb610833799feb497cc87d88c5
</p>

File: types/mongodb/index.d.ts
Comment (HitkoDev) [inline types/mongodb/index.d.ts:2147]: Just like with projection, `sort` should accept input such as `{ 'field.nested': true }`.

**Review by HitkoDev — CHANGES_REQUESTED** [review]

File: types/mongodb/index.d.ts
Comment (Celend) [reply types/mongodb/index.d.ts]: Thanks for the review!

I'm checked documentation and added more type definitions for projection and sort.   

The modified types of projection and sort only for `find` action, so it should have no influence to aggregation.

**Review by HitkoDev — CHANGES_REQUESTED** [review]
Don't forget tests, and make sure the following examples (or equivalent) pass.
### For sort
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a9714077eff8fb4cfa09ebac694174bac02f7f39/types/mongodb/test/cursor.ts#L32

#### E.g.

```ts
.sort({
  score: { 
    $meta: 'textScore'
  },
  age: 1
})
```

and (optionally, probably not possible along with `{ 'field.nested': 1 }` syntax)
```ts
// $ExpectError
.sort({
  notExisting: 1
})
```


### For projection https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a9714077eff8fb4cfa09ebac694174bac02f7f39/types/mongodb/test/cursor.ts#L24

#### E.g.

```ts
.project({
  age: 1,
  addField: 'value',
  renamedAge: '$age',
  softMaxAge: {
    $max: ['$age', 0]
  }
})
```
and (optionally, probably not possible along with `{ 'field.nested': 1 }` syntax)
```ts
// $ExpectError
.project({
  notExisting: 1
})
```

---
Edit: typos, notes

File: types/mongodb/index.d.ts
Comment (HitkoDev) [reply types/mongodb/index.d.ts]: Good job, though there are still some minor changes. While not explicitly documented, when passing sort input as an array of `[key, value]`, value could be any of `1, -1, 'asc', 'desc', 'ascending', 'descending'` in any case, i.e. `'AsC', 'dESC'` https://github.com/mongodb/node-mongodb-native/blob/81684319fe7ed4c89999983d7052ca1029d3f18c/src/utils.ts#L20

I've also added a comment about tests, to make sure these changes don't break more complex projections.

**Comment by Celend** [comment]
> Don't forget tests, and make sure the following examples (or equivalent) pass.
> 
> ### For sort
> https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a9714077eff8fb4cfa09ebac694174bac02f7f39/types/mongodb/test/cursor.ts#L32
> 
> #### E.g.
> ```ts
> .sort({
>   score: { 
>     $meta: 'textScore'
>   },
>   age: 1
> })
> ```
> 
> and (optionally, probably not possible along with `{ 'field.nested': 1 }` syntax)
> 
> ```ts
> // $ExpectError
> .sort({
>   notExisting: 1
> })
> ```
> 
> ### For projection
> https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a9714077eff8fb4cfa09ebac694174bac02f7f39/types/mongodb/test/cursor.ts#L24
> 
> #### E.g.
> ```ts
> .project({
>   age: 1,
>   addField: 'value',
>   renamedAge: '$age',
>   softMaxAge: {
>     $max: ['$age', 0]
>   }
> })
> ```
> 
> and (optionally, probably not possible along with `{ 'field.nested': 1 }` syntax)
> 
> ```ts
> // $ExpectError
> .project({
>   notExisting: 1
> })
> ```
> 
> Edit: typos, notes

Sure! the test is important.  

I searched typescript documentation, seems it's not possible to disallow not existing field but dot notation.

File: types/mongodb/index.d.ts
Comment (avaly) [inline types/mongodb/index.d.ts:1358]:
```suggestion
export type SchemaMember<T, V> = {[P in keyof T]?: V} | {[key: string]: V};
```

Comment (Celend) [reply types/mongodb/index.d.ts]: Oh sorry, the typo fixed, thanks!

File: types/mongodb/index.d.ts
Comment (avaly) [inline types/mongodb/index.d.ts:1123]: nit: This parameter intendation looks off. Either keep all on the same line, or break up each parameter on its own line.

File: types/mongodb/index.d.ts
Comment (avaly) [inline types/mongodb/index.d.ts:1972]: Why is `any` allowed here?

According to https://docs.mongodb.com/manual/reference/method/db.collection.find/#find-projection, only `0`, `1`, `true`, `false` and the `ProjectionOperators` are allowed.

Comment (Celend) [reply types/mongodb/index.d.ts:1972]: That was originally what I thought :D 
But @HitkoDev notice me the MongoDB 4.4 introduced aggregation expression in projection(https://docs.mongodb.com/manual/reference/method/db.collection.find/#find-projection), for example: `{maxInArray: {$max: ['$array', 0]}}`, and it's hard to typing the aggregation expression and maybe some operator not support in projection. 

So for now I added `any` to projection, or have a better solution?

File: types/mongodb/index.d.ts
Comment (Celend) [reply types/mongodb/index.d.ts:1123]: Sure~

File: types/mongodb/tslint.json
Comment (rbuckton) [inline types/mongodb/tslint.json:15]: If this was to work around a build failure due to `@deprecated`, I think that's been fixed in dtslint and these shouldn't be necessary anymore.

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/mongodb/tslint.json
Comment (Celend) [reply types/mongodb/tslint.json:15]: It's still happening :(

How can I fix that?
```
$ npm run lint mongodb 

> definitely-typed@0.0.3 lint /Users/TJT/Projects/DefinitelyTyped
> dtslint types "mongodb"

The 'no-redundant-jsdoc-2' rule threw an error in '/Users/TJT/Projects/DefinitelyTyped/types/mongodb/index.d.ts':
Error: Unexpected tag kind: JSDocDeprecatedTag
    at checkTag (/Users/TJT/Projects/DefinitelyTyped/node_modules/dtslint/bin/rules/noRedundantJsdoc2Rule.js:118:23)
    at cb (/Users/TJT/Projects/DefinitelyTyped/node_modules/dtslint/bin/rules/noRedundantJsdoc2Rule.js:67:25)
    at visitNodes (/Users/TJT/Projects/DefinitelyTyped/node_modules/typescript/lib/typescript.js:27024:30)
    at Object.forEachChild (/Users/TJT/Projects/DefinitelyTyped/node_modules/typescript/lib/typescript.js:27335:21)
    at cb (/Users/TJT/Projects/DefinitelyTyped/node_modules/dtslint/bin/rules/noRedundantJsdoc2Rule.js:72:19)
    at Array.forEach (<anonymous>)
    at walk (/Users/TJT/Projects/DefinitelyTyped/node_modules/dtslint/bin/rules/noRedundantJsdoc2Rule.js:56:34)
    at Rule.AbstractRule.applyWithFunction (/Users/TJT/Projects/DefinitelyTyped/node_modules/tslint/lib/language/rule/abstractRule.js:39:9)
    at Rule.apply (/Users/TJT/Projects/DefinitelyTyped/node_modules/dtslint/bin/rules/noRedundantJsdoc2Rule.js:36:21)
    at Linter.applyRule (/Users/TJT/Projects/DefinitelyTyped/node_modules/tslint/lib/linter.js:214:29)
```

**Review by xamgore — COMMENTED** [review]
Awesome 👍
Could you please bump doc version in changed lines? like `node-mongodb-native/3.1/` → `.../3.6/`

**Comment by Celend** [comment]
> Awesome 👍
> Could you please bump doc version in changed lines? like `node-mongodb-native/3.1/` → `.../3.6/`

Sure, but why `3.6`? should I use the latest version?

File: types/mongodb/test/collection/findX.ts
Comment (avaly) [inline types/mongodb/test/collection/findX.ts:39]: Shouldn't TS complain here about the `notExistingField`?

Comment (Celend) [reply types/mongodb/test/collection/findX.ts:39]: No, because you may use dot notation to access nested object like `obj.prop`, the TS only can treat it as a string, so `notExistingField` is not a valid key but TS can't recognize.

File: types/mongodb/index.d.ts
Comment (andrewbranch) [inline types/mongodb/index.d.ts:1104]: Is every change in this PR correct for 3.5 as well? If not, the version in the header needs to be bumped too.

**Review by andrewbranch — COMMENTED** [review]
I’d like owners to approve before merging this if possible.

File: types/mongodb/index.d.ts
Comment (HitkoDev) [reply types/mongodb/index.d.ts:1104]: MongoDB 4.4 is only _fully_ supported in driver version 3.6. 

Changes in this PR, however, work with driver version 3.5 (and probably even some earlier versions), since the driver just forwards those inputs to the server.

File: types/mongodb/tslint.json
Comment (rbuckton) [reply types/mongodb/tslint.json:15]: Can you check what version of dtslint you have locally via `npm ls dtslint`? The latest version is 3.6.14.

Comment (Celend) [reply types/mongodb/tslint.json]: Oh thanks, the problem fixed!

**Review by HitkoDev — APPROVED** [review]

**Comment by Celend** [comment]
Ready to merge

**Pull request merged**
